kurtosis-devnet/pkg/kurtosis/kurtosis.go

```diff
@@ -0,0 +1,379 @@
+
+		enclaveSpec:      &enclaveSpecAdapter{},
+		enclaveInspecter: &enclaveInspectAdapter{},
+		enclaveObserver:  &enclaveDeployerAdapter{},
+	}
+
+	for _, opt := range opts {
+		opt(d)
+	}
+
+	return d
+}
+
+// templateData holds the data for the command template
+type templateData struct {
+	PackageName string
+	ArgFile     string
+	Enclave     string
+}
+
+// TODO: the following functions follow closely the naming convensions in place
+// with optimism-package. We should probably make them a bit more generic, to
+// make the whole thing less fragile.
+
+// findRPCEndpoint looks for a service matching the given predicate that has an RPC port
+func findRPCEndpoints(services inspect.ServiceMap, matchService func(string) (string, int, bool)) ([]Node, EndpointMap) {
+	interestingPorts := []string{"rpc", "http"}
+	nodeServices := []string{"cl", "el"}
+
+	endpointMap := make(EndpointMap)
+	var nodes []Node
+
+	for serviceName, ports := range services {
+		var port int
+		for _, interestingPort := range interestingPorts {
+			if p, ok := ports[interestingPort]; ok {
+				port = p
+				break
+			}
+		}
+		if port == 0 { // nothing to see here
+			continue
+		}
+
+		if serviceIdentifier, num, ok := matchService(serviceName); ok {
+			var allocated bool
+			for _, service := range nodeServices {
+				if serviceIdentifier == service { // this is a node
+					if num > len(nodes) {
+						nodes = append(nodes, make(Node))
+					}
+					nodes[num-1][serviceIdentifier] = fmt.Sprintf("http://localhost:%d", port)
+					allocated = true
+				}
+			}
+			if !allocated {
+				endpointMap[serviceIdentifier] = fmt.Sprintf("http://localhost:%d", port)
+			}
+		}
+	}
+	return nodes, endpointMap
+}
+
+// return the shorthand service tag (used as key in the final output) and the
+// index if that's a service with multiple instances.
+func serviceTag(serviceName string) (string, int) {
+	// Find index of first number
+	i := strings.IndexFunc(serviceName, func(r rune) bool {
+		return r >= '0' && r <= '9'
+	})
+	if i == -1 {
+		return serviceName, 0
+	}
+	idx, err := strconv.Atoi(serviceName[i : i+1])
+	if err != nil {
+		return serviceName, 0
+	}
+	return serviceName[:i-1], idx
+}
+
+const l2ServiceTagPrefix = "op-"
+
+func findL2Endpoints(services inspect.ServiceMap, suffix string) ([]Node, EndpointMap) {
+	return findRPCEndpoints(services, func(serviceName string) (string, int, bool) {
+		if strings.HasSuffix(serviceName, suffix) {
+			name := strings.TrimSuffix(serviceName, suffix)
+			tag, idx := serviceTag(strings.TrimPrefix(name, l2ServiceTagPrefix))
+			return tag, idx, true
+		}
+		return "", 0, false
+	})
+}
+
+// TODO: L1 services are detected as "non-L2" right now. That might need to change
+// in the future, but for now it's good enough.
+func findL1Endpoints(services inspect.ServiceMap) ([]Node, EndpointMap) {
+	return findRPCEndpoints(services, func(serviceName string) (string, int, bool) {
+		match := !strings.HasPrefix(serviceName, l2ServiceTagPrefix)
```

Comment:
I think the L1 services are prefixed wit `el` and `cl`.

Reply:
yes. I want to make the overall logic around endpoints identification / tagging more readable. I'll  refactor all that in a followup